### PR TITLE
charset: support builtin function convert

### PR DESCRIPTION
**Pull request by xiongjiwei**
### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/28657

- if convert to different charset, we just need to check the character is convertible or not, use `?` to replace the convertible character
- convert function should always derive to `CoercibilityImplicit`, see the new tests

### Check List 

Tests 

- [X] Unit test
- [X] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note 

```release-note
Please add a release note, or a 'None' if it is not needed.
```

### CODE REVIEW

**Comment by Defined2014** [comment]
LGTM

**Review by Defined2014 — APPROVED** [review]

File: parser/charset/encoding.go [hunk 1]
```diff
@@ -99,6 +99,36 @@ func (e *Encoding) Encode(dest, src []byte) ([]byte, error) {
...
+func (e *Encoding) EncodeInternal(dest, src []byte) []byte {
+	if !e.Enabled() {
+		return src
+	}
+	if dest == nil {
+		dest = make([]byte, 0, len(src))
+	}
+	var srcOffset int
+
+	buf := make([]byte, 4)
```
Comment (tangenta) [inline parser/charset/encoding.go:116]:
```suggestion
	var buf [4]byte
```

Comment (tangenta) [reply parser/charset/encoding.go:116]: We can avoid the allocation from heap.

**Review by zimulala — APPROVED** [review]
LGTM

File: parser/charset/encoding.go (hunk 1 above)
Comment (xiongjiwei) [reply parser/charset/encoding.go:127]: no, we cannot, [4]byte is different from []byte.

**Review by tangenta — APPROVED** [review]

**Pull request merged**